### PR TITLE
Update operations return leader info when leader change event happens

### PR DESCRIPTION
**Pull request by zlcook**
As title

### CODE REVIEW

**Comment by zlcook** [comment]
Jenkins go

File: src/storage/UpdateEdgeProcessor.cpp [hunk 1]
```diff
@@ -364,10 +364,22 @@ void UpdateEdgeProcessor::process(const cpp2::UpdateEdgeRequest& req) {
...
         },
         [this, partId, edgeKey, req] (kvstore::ResultCode code) {
-            this->pushResultCode(this->to(code), partId);
+            cpp2::ResultCode thriftResult;
+            thriftResult.set_code(to(code));
+            thriftResult.set_part_id(partId);
+            if (code == kvstore::ResultCode::ERR_LEADER_CHANGED) {
+                nebula::cpp2::HostAddr leader;
+                auto addrRet = kvstore_->partLeader(this->spaceId_, partId);
+                CHECK(ok(addrRet));
```
Comment (monadbobo) [inline src/storage/UpdateEdgeProcessor.cpp]: Here should return failure.

File: src/storage/UpdateVertexProcessor.cpp [hunk 2]
```diff
@@ -341,10 +341,22 @@ void UpdateVertexProcessor::process(const cpp2::UpdateVertexRequest& req) {
...
         },
         [this, partId, vId, req] (kvstore::ResultCode code) {
-            this->pushResultCode(this->to(code), partId);
+            cpp2::ResultCode thriftResult;
+            thriftResult.set_code(to(code));
+            thriftResult.set_part_id(partId);
+            if (code == kvstore::ResultCode::ERR_LEADER_CHANGED) {
+                nebula::cpp2::HostAddr leader;
+                auto addrRet = kvstore_->partLeader(this->spaceId_, partId);
+                CHECK(ok(addrRet));
```
Comment (monadbobo) [inline src/storage/UpdateVertexProcessor.cpp]: the same as above

**Comment by critical27** [comment]
Maybe we can reuse the method `pushResultCode(cpp2::ErrorCode code, PartitionID partId, HostAddr leader)` in BaseProcessor.

File: src/storage/UpdateVertexProcessor.cpp [hunk 3]
```diff
@@ -341,10 +341,22 @@ void UpdateVertexProcessor::process(const cpp2::UpdateVertexRequest& req) {
...
+                auto addr = value(std::move(addrRet));
+                leader.set_ip(addr.first);
+                leader.set_port(addr.second);
+                thriftResult.set_leader(std::move(leader));
+            }
             if (code == kvstore::ResultCode::SUCCEEDED) {
                 onProcessFinished(req.get_return_columns().size());
             } else {
+                this->codes_.emplace_back(std::move(thriftResult));
```
Comment (Shylock-Hg) [inline src/storage/UpdateVertexProcessor.cpp]: Why not use `pushResultCode`.

Comment (Shylock-Hg) [reply src/storage/UpdateVertexProcessor.cpp]: And better use `if ... else if ... else` statements.

Comment (zlcook) [reply src/storage/UpdateVertexProcessor.cpp]: Ok

File: src/storage/UpdateEdgeProcessor.cpp (hunk 1 above)
Comment (zlcook) [reply src/storage/UpdateEdgeProcessor.cpp]: > Here should return failure.

Why?

Comment (monadbobo) [reply src/storage/UpdateEdgeProcessor.cpp]: We can't crash the process at will, we must handle the errors that can be handled.

PS: What I mean here is that you need to determine whether addrRet is legal.

Comment (zlcook) [reply src/storage/UpdateEdgeProcessor.cpp]: > We can't crash the process at will, we must handle the errors that can be handled.
> 
> PS: What I mean here is that you need to determine whether addrRet is legal.

You are right

**Comment by zlcook** [comment]
Jenkins go

File: src/storage/UpdateVertexProcessor.cpp [hunk 4]
```diff
@@ -341,12 +341,25 @@ void UpdateVertexProcessor::process(const cpp2::UpdateVertexRequest& req) {
...
+                    auto addrRet = kvstore_->partLeader(this->spaceId_, partId);
+                    if (ok(addrRet)) {
+                        auto leader = value(std::move(addrRet));
+                        this->pushResultCode(to(code), partId, leader);
+                        break;
+                    }
+                    LOG(ERROR) << "Fail to get part leader, spaceId: " << this->spaceId_
+                               << ", partId: " << partId << ", ResultCode: " << error(addrRet);
+                }
+                this->pushResultCode(to(code), partId);
```
Comment (yixinglu) [inline src/storage/UpdateVertexProcessor.cpp]: Why not extract these same logics to a function like `handleLeaderChanged`?

Comment (zlcook) [reply src/storage/UpdateVertexProcessor.cpp]: > Why not extract these same logics to a function like `handleLeaderChanged` in base class?

No reason to refuse you...

**Comment by zlcook** [comment]
Jenkins Go

**Comment by jude-zhu** [comment]
Jenkins Go

File: src/storage/BaseProcessor.h [hunk 5]
```diff
@@ -98,6 +98,17 @@ class BaseProcessor {
         }
     }
 
+    void handleLeaderChanged(GraphSpaceID spaceId, PartitionID partId) {
+        auto addrRet = kvstore_->partLeader(spaceId, partId);
+        if (ok(addrRet)) {
+            auto leader = value(std::move(addrRet));
+            this->pushResultCode(cpp2::ErrorCode::E_LEADER_CHANGED, partId, leader);
+        } else {
+            LOG(ERROR) << "Fail to get part leader, spaceId: " << spaceId
```
Comment (dangleptr) [inline src/storage/BaseProcessor.h]: Be careful.  If we can't get the leader information (which is normal),  we should return the empty leader information and errorcode Leader_change. 

By the way,  could we unify the duplicate code about leader change logic with handleAsync in BaseProcessor

Comment (zlcook) [reply src/storage/BaseProcessor.h]: Oh my fault

File: src/storage/BaseProcessor.h [hunk 6]
```diff
@@ -98,6 +98,18 @@ class BaseProcessor {
...
+                       << ", partId: " << partId << ", ResultCode: " << error(addrRet);
+            this->pushResultCode(cpp2::ErrorCode::E_LEADER_CHANGED, partId);
```
Comment (dangleptr) [inline src/storage/BaseProcessor.h]: Return the real error code (error(addrRet)) instead of E_LEADER_CHANGED here.

Comment (zlcook) [reply src/storage/BaseProcessor.h]: > Return the real error code (error(addrRet)) instead of E_LEADER_CHANGED here.
Good catched

**Comment by zlcook** [comment]
Everybody go on

**Pull request merged**